rest.go

```diff
@@ -90,39 +51,39 @@ func BuildResponse(res *http.Response) (*Response, error) {
 }
 
 // API is the main interface to the API.
-func API(request Request) (*Response, error) {
+func API(request *http.Request) (*Response, error) {
 	return DefaultClient.API(request)
 }
 
 // The following functions enable the ability to define a
 // custom HTTP Client
 
 // MakeRequest makes the API call.
-func (c *Client) MakeRequest(req *http.Request) (*http.Response, error) {
+func (c *Client) makeRequest(req *http.Request) (*http.Response, error) {
 	return c.HTTPClient.Do(req)
 }
 
 // API is the main interface to the API.
-func (c *Client) API(request Request) (*Response, error) {
-	// Add any query parameters to the URL.
-	if len(request.QueryParams) != 0 {
-		request.BaseURL = AddQueryParameters(request.BaseURL, request.QueryParams)
+func (c *Client) API(request *http.Request) (*Response, error) {
+	if request == nil {
+		return nil, errors.New("invalid request: request cannot be nil")
 	}
+	_, supportedMethod := supportedMethods[request.Method]
 
-	// Build the HTTP request object.
-	req, err := BuildRequestObject(request)
-	if err != nil {
-		return nil, err
+	if !supportedMethod {
+		return nil, errors.New("unsupported request method:" + request.Method)
+	}
```

Comment:
I'm sorry I probably wasn't very clear. What I'm referring to specifically is this particular idiom here https://github.com/sendgrid/rest/blob/master/rest.go#L12. Because there was a `type Method string` previously then inherently because of the beauty of Go's type checking it was impossible to issue a request using an unsupported method. Alternatively we can allow all manner of requests to be submitted and just error for unsupported ones through the http semantics we'd get for free from the sendgrid service that underlies this client library I'm cool with that as well. What do you think?